crates/apollo-compiler/src/validation/fragment.rs

```diff
@@ -1,8 +1,175 @@
+        .get_possible_types(parent_type.clone())
+        .into_iter()
+        .map(|ty| ty.name().to_string())
+        .collect::<HashSet<_>>();
+    let concrete_condition_types = db
+        .get_possible_types(cond_type)
+        .into_iter()
+        .map(|ty| ty.name().to_string())
+        .collect::<HashSet<_>>();
+
+    let mut applicable_types = concrete_parent_types.intersection(&concrete_condition_types);
+    if applicable_types.next().is_none() {
+        // Report specific errors for the different kinds of fragments.
+        let diagnostic = match spread {
+            hir::FragmentSelection::FragmentSpread(spread) => {
+                // This unwrap is safe because the fragment definition must exist for `cond_type` to be `Some()`.
+                let fragment_definition = spread.fragment(db.upcast()).unwrap();
+
+                ApolloDiagnostic::new(
+                    db,
+                    spread.loc().into(),
+                    DiagnosticData::InvalidFragmentSpread {
+                        name: Some(spread.name().to_string()),
+                        type_name: parent_type.name().to_string(),
+                    },
+                )
+                .label(Label::new(
+                    spread.loc(),
+                    format!("fragment `{}` cannot be applied", spread.name()),
+                ))
+                .label(Label::new(
+                    fragment_definition.loc(),
+                    format!("fragment declared with type condition `{cond}` here"),
+                ))
+                .label(Label::new(
+                    parent_type.loc(),
+                    format!("type condition `{cond}` is not assignable to this type"),
+                ))
+            }
+            hir::FragmentSelection::InlineFragment(inline) => ApolloDiagnostic::new(
+                db,
+                inline.loc().into(),
+                DiagnosticData::InvalidFragmentSpread {
+                    name: None,
+                    type_name: parent_type.name().to_string(),
+                },
+            )
+            .label(Label::new(
+                inline.loc(),
+                format!("fragment applied with type condition `{cond}` here"),
+            ))
+            .label(Label::new(
+                parent_type.loc(),
+                format!("type condition `{cond}` is not assignable to this type"),
+            )),
+        };
+
+        diagnostics.push(diagnostic);
+    }
+
+    diagnostics
+}
+
+pub fn validate_inline_fragment(
+    db: &dyn ValidationDatabase,
+    inline: Arc<hir::InlineFragment>,
+) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    diagnostics.extend(db.validate_directives(
+        inline.directives().to_vec(),
+        hir::DirectiveLocation::InlineFragment,
+    ));
+
+    diagnostics.extend(db.validate_fragment_type_condition(
+        inline.type_condition().map(|t| t.to_string()),
+        inline.loc(),
+    ));
```

Comment:
i wonder if perhaps some of the "double" reporting can be avoided if `validate_fragment_type_condition` returns a result. we can then only proceed to `validate_fragment_selection` if the previous validation is `Ok`?

Reply:
hmmm thats worth a shot 🤔

Reply:
did this by checking `validate_fragment_type_condition().is_empty()